File: leetcode/easy/876/876.js
/** 
 * Runtime: 72 ms, faster than 88.62% 
 * Memory Usage: 38.3 MB, less than 60.49%
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var middleNode = function(head) {
  let size = 0;

  let temp = head;
  while(temp) {
    size+=1;
    temp = temp.next;
  }

  size >>= 1;
  temp = head;
  for(let i=0; i<size; ++i) {
    temp = temp.next;
  }

  return temp;
};

File: leetcode/easy/876/876.rb
# Runtime: 76 ms, faster than 11.36%
# Memory Usage: 209.8 MB, less than 50.00%

# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @return {ListNode}
def middle_node(head)
  cnt = 0

  temp = head
  while temp
    cnt += 1
    temp = temp.next 
  end

  cnt >>= 1
  temp = head
  for i in (0...cnt) do
    temp = temp.next 
  end

  temp
end
